_posts/2019-03-19-til-validations-and-constraints.md

```diff
@@ -0,0 +1,229 @@
+---
+author: Prince Wilson
+date: 2019-03-19
+layout: post
+categories: til
+author_link: https://github.com/maxcell
+title:  TIL Constraints and Validations
+excerpt: >
+  Let's take a look at how Ecto handles these two ways
+  of ensuring data integrity
+---
+
+Developers want to create the best applications they can for their users.
+In the process, they want to make sure to give good feedback to their users
+when data doesn't get saved into the database. In Elixir, there is a great tool
+on top of the database that helps, Ecto! It can put validations and
+constraints onto specific fields to ensure data integrity.
+
+However, did you know they are differences between validations and constraints?
+I didn't. In the process of building a side project, I ran into the problem a few
+times! Let's talk about the goal of them and see the differences. We'll dive into
+**why we need both each** towards the end.
+
+## Data Integrity is Rule #1
+> Data Integrity is the maintenance of, and the assurance of the accuracy and
+> consistency of, data over its entire lifecycle
+> - [Wikipedia](https://en.wikipedia.org/wiki/Data_integrity)
+
+So we're building a super cool app with users that can login and logout.
+We'd probably have some schema like this:
+```elixir
+# Using Phoenix 1.4 with Contexts but still applies all the same
+defmodule MyCoolWebApp.Accounts.User do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  schema "users" do
+    field :display_name, :string
+    field :email, :string
+    field :password_hash, :string
+    field :password, :string, virtual: true
+
+    timestamps()
+  end
+end
+```
+
+And inside of it, we'd want to describe the `changeset`:
```

Comment:
Could we add a super short line on what a changeset is? Something like "Ecto changesets allow us to cast, filter, validate and constrain the structs created by our schema that represent our database records". Then you can show your first, simple changeset that _only_ casts and explain the drawback you laid out.